File: src/extension/src/scoring/ai-style-scorer.ts
import { aiStylePatterns } from '../masters';
import { patternDetectors } from './pattern-detectors';
import { computeTextStats } from './text-stats';
import type { ScoreReason, ScoreResult } from './types';

const MIN_JUDGEABLE_BODY_LENGTH = 200;
const SCORE_MIN = 0;
const SCORE_MAX = 100;

export type AiStyleScoreResult = ScoreResult | 'unjudgeable';

export class AiStyleScorer {
  constructor(private readonly body: string) {}

  score(): AiStyleScoreResult {
    if (this.body.length < MIN_JUDGEABLE_BODY_LENGTH) {
      return 'unjudgeable';
    }

    const stats = computeTextStats(this.body);
    const reasons: ScoreReason[] = [];

    for (const pattern of aiStylePatterns) {
      const detector = patternDetectors[pattern.pattern_code];
      if (detector === undefined) {
        throw new Error(`未対応のAI文体特徴パターンです: ${pattern.pattern_code}`);
      }
      if (detector(stats)) {
        reasons.push({ code: pattern.pattern_code.toUpperCase(), delta: pattern.weight });
      }
    }

    const rawScore = reasons.reduce((total, reason) => total + reason.delta, 0);
    const score = Math.min(SCORE_MAX, Math.max(SCORE_MIN, rawScore));
    return { score, reasons };
  }
}

File: src/api/app/services/rule_based/ai_style_scorer.rb
# src/extension/src/scoring/ai-style-scorer.ts のRuby移植版。
# requirements.md 1.4-B準拠。
module RuleBased
  class AiStyleScorer
    class UnsupportedPatternError < StandardError; end

    MIN_JUDGEABLE_BODY_LENGTH = 200
    SCORE_MIN = 0
    SCORE_MAX = 100
    UNJUDGEABLE = :unjudgeable

    def initialize(body)
      @body = body
    end

    def score
      return UNJUDGEABLE if @body.length < MIN_JUDGEABLE_BODY_LENGTH

      stats = TextStatsCalculator.compute(@body)
      reasons = AiStylePattern.find_each.filter_map do |pattern|
        detector = PatternDetectors::DETECTORS[pattern.pattern_code]
        if detector.nil?
          raise UnsupportedPatternError, "未対応のAI文体特徴パターンです: #{pattern.pattern_code}"
        end

        { code: pattern.pattern_code.upcase, delta: pattern.weight } if detector.call(stats)
      end

      raw_score = reasons.sum { |r| r[:delta] }
      { score: raw_score.clamp(SCORE_MIN, SCORE_MAX), reasons: reasons }
    end
  end
end
